### ISSUE TITLE
Include intersecting polyhedra in Polyhedral_mesh_domain_with_features_3

### ISSUE THREAD

**Issue by matekelemen** [CGAL/cgal#4814]
## Issue Details

A simple example: create a 3D mesh of a cube with its midplane protected.

I opened a [question on stackoverflow](https://stackoverflow.com/questions/62451016/cgal-detect-features-creates-unprotectable-features/62459032?noredirect=1#comment110487277_62459032) about this a while ago but still couldn't solve this problem adequately. I'm trying to follow the [Remeshing a Polyhedral Domain with Surfaces](https://doc.cgal.org/latest/Mesh_3/Mesh_3_2mesh_polyhedral_domain_with_surface_inside_8cpp-example.html) example, but the obvious problem is that the edges of the cube and the midplane intersect.

### What I tried so far

Adding the the midplane as polyhedral surface and the cube as the bounding domain results in ```make_mesh_3``` failing with ```minimal_size_ > 0 || sq_d > 0```. I tried merging the midplane and cube polyhedra (which is a pain in itself, but more on this later), but I still need an intersection-free bounding surface in this case. 

My most desperate attempt was to merge the midplane and the cube polyhedra, then create a slightly extended bounding box around it, and use that as the bounding domain. After meshing, I delete every cell from the triangulation that lies outside the original cube (the code for this is rather lengthy so I'd avoid posting it here if possible, but I can provide it if it's absolutely necessary). Although this works, it takes a surprisingly long time to mesh (~10-15 seconds). However, for real-world problems, the meshing never finishes and I have a suspicion why:

The cube+midplane polygon soup is non-manifold, so when [polygon_soup_to_polygon_mesh](https://doc.cgal.org/latest/Polygon_mesh_processing/group__PMP__repairing__grp.html#gae4bb69cfd25240e8f929239b02d8c7f8) is called on it to create a ```Mesh_polyhedron_3```, a lot of faces and their vertices are duplicated, and I'm guessing this complicates things for ```make_mesh_3```.

### Additional remarks

I should probably note that the surfaces I want to protect never clip through each other. At most, they have overlapping faces, or vertices intersecting edges (just like the cube with midplane example).

I couldn't find any built-in function for merging ```Polyhedron_3```s. I could probably make it work with [corefine_and_compute_boolean_operations](https://doc.cgal.org/latest/Polygon_mesh_processing/group__PMP__corefinement__grp.html#ga70346bbfd568de05220767cdcbcd6f68) but the surfaces don't always bound a volume so that's out of the question. Any suggestions how to do this with high-level CGAL functions?

When calling ```polygon_soup_to_polygon_mesh``` with the mesh type being ```Surface_mesh```, no vertices or faces are duplicated. However, if I convert it to ```Mesh_polyhedron_3``` with [copy_face_graph](https://doc.cgal.org/latest/BGL/group__PkgBGLHelperFct.html#ga2d3eb044bfe42eab566779ce32216c11), the same vertices and faces get duplicated as before.

## Source Code

The toy example from [stackoverflow](https://stackoverflow.com/questions/62451016/cgal-detect-features-creates-unprotectable-features/62459032?noredirect=1#comment110487277_62459032):
```
// --- External Includes ---
#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
#include <CGAL/Mesh_polyhedron_3.h>

#include <CGAL/Polyhedral_mesh_domain_with_features_3.h>
#include <CGAL/Mesh_triangulation_3.h>
#include <CGAL/Mesh_complex_3_in_triangulation_3.h>
#include <CGAL/Mesh_criteria_3.h>
#include <CGAL/make_mesh_3.h>

// CGAL types
namespace cgal {

using Kernel            = CGAL::Exact_predicates_inexact_constructions_kernel;
using ConcurrencyTag    = CGAL::Sequential_tag;
using PolyhedronSurface = CGAL::Mesh_polyhedron_3<Kernel>::type;

using MeshDomain        = CGAL::Polyhedral_mesh_domain_with_features_3<Kernel>;
using Tr                = CGAL::Mesh_triangulation_3<MeshDomain,CGAL::Default,ConcurrencyTag>::type;
using Triangulation     = CGAL::Mesh_complex_3_in_triangulation_3<Tr>;
using MeshCriteria      = CGAL::Mesh_criteria_3<Tr>;

using Point             = cgal::MeshDomain::Point_3;
}



int main()
{
    // Define points for the cube and midPlane
    std::vector<cgal::Point> points = 
    {
        cgal::Point(  1.0,    0.0,    0.0),
        cgal::Point(  1.0,    1.0,    0.0),
        cgal::Point(  0.0,    1.0,    0.0),
        cgal::Point(  0.0,    0.0,    0.0),
        cgal::Point(  1.0,    0.0,    1.0),
        cgal::Point(  1.0,    1.0,    1.0),
        cgal::Point(  0.0,    1.0,    1.0),
        cgal::Point(  0.0,    0.0,    1.0),

        cgal::Point(  1.0,    0.0,    0.5),
        cgal::Point(  1.0,    1.0,    0.5),
        cgal::Point(  0.0,    1.0,    0.5),
        cgal::Point(  0.0,    0.0,    0.5)
    };

    // Create polyhedra
    cgal::PolyhedronSurface cube, midPlane;
    cube.make_triangle( points[0],  points[3],  points[1]);
    cube.make_triangle( points[1],  points[3],  points[2]);
    cube.make_triangle( points[4],  points[5],  points[7]);
    cube.make_triangle( points[5],  points[6],  points[7]);

    cube.make_triangle( points[0],  points[4],  points[3]);
    cube.make_triangle( points[3],  points[4],  points[7]);
    cube.make_triangle( points[1],  points[2],  points[5]);
    cube.make_triangle( points[2],  points[6],  points[5]);

    cube.make_triangle( points[2],  points[3],  points[6]);
    cube.make_triangle( points[3],  points[7],  points[6]);
    cube.make_triangle( points[0],  points[1],  points[5]);
    cube.make_triangle( points[0],  points[5],  points[4]);

    midPlane.make_triangle( points[8],  points[9],  points[10]);
    midPlane.make_triangle( points[8],  points[10], points[11]);

    // Triangulation
    cgal::MeshDomain meshDomain( midPlane, cube );
    meshDomain.detect_features();

    cgal::MeshCriteria meshCriteria(    CGAL::parameters::facet_angle   = 30,
                                        CGAL::parameters::edge_size     = 0.2 );

    auto triangulation = CGAL::make_mesh_3<cgal::Triangulation>(    meshDomain,
                                                                    meshCriteria );

    return 0;
}
```

### Questions

In descending order of importance:
* Is it possible to protect intersecting polyhedral surfaces in a domain?
* Is it possible to merge ```Polyhedron_3```s easily with built-in CGAL functions?
* Is there a way to convert a ```Surface_mesh``` to a ```Polyhedron_3``` without duplicating any vertices or faces?
* Are duplicate vertices and faces a problem for ```make_mesh_3```?
* What can be the reason for ```make_mesh_3``` taking unreasonably long even for simple geometries?

## Environment

* Operating system: Ubuntu 20.04 LTS
* Compiler: gcc-10.1
* Debug
* CGAL version: Latest from source
* Boost version: Latest from source

**Comment by sloriot** [comment]
In this particular case, you should [`corefine()`](https://doc.cgal.org/latest/Polygon_mesh_processing/group__PMP__corefinement__grp.html#ga6447dee822aaf92016f34512ce0b3456) the cube and use the `edge_is_constrained_map()` named parameter to collect intersection edges. Then you can split the cube in two parts using the function [`split_connected_components()`](https://cgal.geometryfactory.com/CGAL/doc/master/Polygon_mesh_processing/group__keep__connected__components__grp.html#ga9ddd1e4b915a4232b1ce5611985302aa) available in CGAL 5.1-beta by also using the `edge_is_constrained_map()` named parameter.
Then you provide your three surfaces to the mesh domain [`Polyhedral_mesh_domain_3`](https://doc.cgal.org/latest/Mesh_3/classCGAL_1_1Polyhedral__mesh__domain__3.html) or [`Polyhedral_mesh_domain_with_features_3`](https://doc.cgal.org/latest/Mesh_3/classCGAL_1_1Polyhedral__mesh__domain__with__features__3.html) with the correct labels and it should work.

**Comment by sloriot** [comment]
I'm closing this issue feel free to reopen it if you think the answer was not sufficent to solve your problem.
